### ISSUE TITLE
[-Wunsafe-buffer-usage]: false negative on smart pointers

### ISSUE THREAD

**Issue by juanvazquez** [llvm/llvm-project#73452]
C++ Smart Pointers (for example std::unique_ptr) have an `operator[]` that provides access to elements of an array managed by the smart pointer.

As far as I can tell, the `operator[]` for smart pointers cannot be hardened by the libc++ because the smart pointer doesn't store size information about the underlying array.

Hence it would be convenient have accesses flagged with `-Wunsafe-buffer-usage`. But that's not the case today. 

Example:

```cpp
#include <memory>

void foo() {
  std::unique_ptr<char[]> ptr = std::make_unique<char[]>(16);
  ptr[0]; // -Wunsafe-buffer-usage does not warn about it
}

void bar() {
  char *arr = new char[5];
  arr[4]; // -Wunsafe-buffer-usage does warns about it
}

int main() {
  bar();
  foo();
  return 0;
}
```

Running `-Wunsafe-buffer-usage` in the example correctly warns the function bar(), but doesn't emit any warning on the function foo():

```
/tmp/test.cc:10:3: warning: unsafe buffer access [-Wunsafe-buffer-usage]
   10 |   arr[4];
      |   ^~~
/tmp/test.cc:10:3: note: pass -fsafe-buffer-usage-suggestions to receive code hardening suggestions
```

I think that accessing the elements via `std::unique_ptr::operator[]` emits a `CXXOperatorCallExpr` instead of a `ArraySubscriptExpr`, hence the `-Wunsafe-buffer-usage` analysis do not flag them.

I think would be convenient to flag accesses to memory via smart pointers via `operator[]`, that could be out of bounds. What do you think? Any feedback?

**Comment by juanvazquez** [comment]
Hi @malavikasamak @ziqingluo-90  @t-rasmud @jkorous-apple 

In the Safe Buffers doc (clang/docs/SafeBuffers.rst) I read

> The compiler warning -Wunsafe-buffer-usage is built to assist you with this step of the process. A -Wunsafe-buffer-usage warning is emitted whenever one of the following buffer operations are performed on a raw pointer:
> 
> ...
> * C++ smart pointer operations such as std::unique_ptr<T[N]>::operator[](), which unfortunately cannot be made fully safe within the rules of the C++ standard (as of C++23).

Did a quick test and afaict -Wunsafe-buffer-usage doesn't warns on std::unique_ptr<T[N]>::operator[]() as of today. 

I was asking myself if I missed something or that supports is in the works? If the idea is planned but you lack time, would be a patch be welcomed?

**Comment by jkorous-apple** [comment]
Hi @juanvazquez

You are right that the warning doesn't report pointer arithmetic with standard library smart pointers yet.
A patch would be welcome. We should add a new subgroup of `-Wunsafe-buffer-usage` just like we did for `-Wunsafe-buffer-usage-in-container` or `-Wunsafe-buffer-usage-in-libc-call`. That way the new cases can be rolled out incrementally.

This topic has been brought up just last week on LLVM Discord and a patch might be already in the works.
We should coordinate:
https://discord.com/channels/636084430946959380/1345431477424492625/1384706511867220092

**Comment by ldionne** [comment]
Just to clarify, `std::unique_ptr<T[N]>` does not provide an `operator[]`. However, `std::unique_ptr<T[]>` does provide `operator[]` and IMO `-Wunsafe-buffer-usage` should probably warn on it.

That being said, libc++ also provides an ABI configuration under which `std::unique_ptr<T[]>::operator[]` is hardened, in which case the warning should be silenced. Perhaps this means that the best way to achieve this whole thing is to patch libc++ instead by adding the `__unsafe_buffer_usage__` attribute to `std::unique_ptr<T[]>::operator[]`, but only when it's not hardened?

**Comment by juanvazquez** [comment]
@jkorous-apple , thanks for the update. If a patch is in progress, that works. Looking forward to it.

@ldionne 

> Perhaps this means that the best way to achieve this whole thing is to patch libc++ instead by adding the __unsafe_buffer_usage__ attribute to std::unique_ptr<T[]>::operator[], but only when it's not hardened?

imho, that sounds like it could be made.

**Comment by jkorous-apple** [comment]
Considering the choice between hardcoding this in `-Wunsafe-buffer-usage` analysis in clang or adding the attribute to libc++ I realized I don't have clarity about the programming model itself. Since the model might imply implementation details why don't we talk about it first?

IMO `-Wunsafe-buffer-usage` should NOT report use of `std::unique_ptr<T[]>::operator[]` only if it is guaranteed to do the correct runtime bounds check.

I am trying to understand what guarantees does unique_ptr provide from the hardening description:
https://github.com/llvm/llvm-project/blob/main/libcxx/include/__memory/unique_ptr.h#L300

Let's start with the case when `_LIBCPP_ABI_BOUNDED_UNIQUE_PTR` is enabled.

When `unique_ptr<T[]>` is constructed from a raw pointer the newly constructed instance doesn't receive any bounds information. For such instances we'd have to emit `-Wunsafe-buffer-usage` every time `operator[]` is used because it won't be bounds-safe. Because we can't relate every `operator[]` call with a construction site at compile time (we can't afford advanced/slow analysis at compilation) we would still have to warn about every `operator[]` call.

I wonder if we can improve this by diagnosing unsafe `unique_ptr` construction instead. Can we declare such constructors that don't propagate bounds as unsafe and diagnose their use at compile-time?
For special patterns like `unique_ptr(new int[42])` the users could use `make_unique` (and we could emit a fixit in clang to guide them).
We probably could set the bound to zero in the implementation of `unique_ptr(nullptr)`.
In the general case (`unique_ptr<int> foo{a_pointer};`) the users would probably have to change the initializer to `unique_ptr` (possibly recursively).

Similarly we'd probably have to prohibit and diagnose use of `unique_ptr::reset(T*)`.

If we do that - is it true that if all `-Wunsafe-buffer-usage` warnings are addressed (and not suppressed) that all calls to `std::unique_ptr<T[]>::operator[]` will have access to bounds information and can do a bounds check at runtime?

**Comment by juanvazquez** [comment]
@jkorous-apple, for constructions/assignments/reset/swap from an existing ptr/allocation, wouldn't you have to track related ptrs still? 

If relating `operator[]` to its declaration is too expensive, won't be relating different pointers expensive too?

Or what you are proposing is just consider construction/assignment/reset/swap from existing pointers as unsafe?

**Comment by jkorous-apple** [comment]
@juanvazquez I am proposing labeling any `std::unique_ptr` method that can leave its object with unknown or invalid bound as unsafe. I don't think we can consider any use of `std::unique_ptr::operator[]` as safe otherwise.

I am also saying that if we do that and assume that there are no such unsafe operations in the code then `std::unique_ptr::operator[]` will always have correct bound information and in the hardened mode will always do the correct runtime check.

**Comment by juanvazquez** [comment]
@jkorous-apple , I see. that maybe a reasonable approach when `_LIBCPP_ABI_BOUNDED_UNIQUE_PTR` is in place. 

And then, the idea would be to, otherwise, warn as far as the managed type does not `__has_array_cookie`? And on this case, warn on `unique_ptr::operator[]`?

**Comment by jkorous-apple** [comment]
@juanvazquez IIUC only buffers on heap would have the cookie. Is that right?

What if someone constructs unique_ptr from a pointer to a stack variable or a global?

Also, it would be only the first element in the buffer that is preceded by the cookie.
What if someone accidentally creates unique_ptr from the address of second element in the buffer?

Neither of these make much sense to do but the programming model and the diagnostics should guarantee bounds-safety even in such cases.

WDYT?

**Comment by juanvazquez** [comment]
@jkorous-apple I do think the hardening story of unique_ptr, as is today, while smart, is kinda convoluted, and maybe would be a better approach to just encourage migration to C++ hardened containers, as part of the safe buffers programming model.

But I do think your remarks on unique_ptr are valid.

**Comment by jkorous-apple** [comment]
@juanvazquez Do you mean that `unique_ptr<T[]>` should not be used at all and the replacement would be for example `unique_ptr<vector<T>>`?

**Comment by juanvazquez** [comment]
@jkorous-apple , not exactly. I do think that `unique_ptr<T[]>` could be treat similarly to raw pointers `T*` to sequences, and hence being migrated to `vector<T>`, or other container, `vector` may not be always the adequate container, the same way that `vector` is not always the better fit when migrating sequences pointed by `T*`.

**Comment by jkorous-apple** [comment]
Got it.
Are you saying that the guarantees `unique_ptr` hardening provides are hard to reason about and we should just emit `-Wunsafe-buffer-usage` warnign for any `unique_ptr::operator[]`?

**Comment by juanvazquez** [comment]
I do think so, yes. And maybe reconsider if the hardening of smart pointers, as it is, is the correct tool to mitigate spatial safety.

But I do understand that upstream leans towards hardened smart pointers atm. In such case, I do understand it's preferable to have the warning accounting for the different situations. Otherwise, the user is left with the mission of having to reason by himself if the smart pointer is safe to access a sequence or not on a case by case basis.

**Comment by jkorous-apple** [comment]
For now emitting the warning for any `unique_ptr::operator[]` use-site and hardcoding it in `-Wunsafe-buffer-usage` analysis sounds reasonable to me.

When there is a libc++ configuration that provides guarantees about unique_ptr bounds-safety we can switch to the attribute-based approach.
